vault/identity_store_oidc_provider.go

```diff
@@ -0,0 +1,164 @@
+			HelpDescription: "Create, Read, Update, and Delete OIDC assignments.",
+		},
+		{
+			Pattern: "oidc/assignment/?$",
+			Operations: map[logical.Operation]framework.OperationHandler{
+				logical.ListOperation: &framework.PathOperation{
+					Callback: i.pathOIDCListAssignment,
+				},
+			},
+			HelpSynopsis:    "List OIDC assignments",
+			HelpDescription: "List all configured OIDC assignments in the identity backend.",
+		},
+	}
+}
+
+// pathOIDCCreateUpdateAssignment is used to create a new assignment or update an existing one
+func (i *IdentityStore) pathOIDCCreateUpdateAssignment(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	name := d.Get("name").(string)
+
+	var assignment assignment
+	if req.Operation == logical.UpdateOperation {
+		entry, err := req.Storage.Get(ctx, assignmentPath+name)
+		if err != nil {
+			return nil, err
+		}
+		if entry != nil {
+			if err := entry.DecodeJSON(&assignment); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	if entitiesRaw, ok := d.GetOk("entities"); ok {
+		assignment.Entities = entitiesRaw.([]string)
+	} else if req.Operation == logical.CreateOperation {
+		assignment.Entities = d.GetDefaultOrZero("entities").([]string)
+	}
+
+	if groupsRaw, ok := d.GetOk("groups"); ok {
+		assignment.Groups = groupsRaw.([]string)
+	} else if req.Operation == logical.CreateOperation {
+		assignment.Groups = d.GetDefaultOrZero("groups").([]string)
+	}
+
+	// store assignment
+	entry, err := logical.StorageEntryJSON(assignmentPath+name, assignment)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := req.Storage.Put(ctx, entry); err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+// pathOIDCListAssignment is used to list assignments
+func (i *IdentityStore) pathOIDCListAssignment(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	assignments, err := req.Storage.List(ctx, assignmentPath)
+	if err != nil {
+		return nil, err
+	}
+	return logical.ListResponse(assignments), nil
+}
+
+// pathOIDCReadAssignment is used to read an existing assignment
+func (i *IdentityStore) pathOIDCReadAssignment(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	name := d.Get("name").(string)
+
+	entry, err := req.Storage.Get(ctx, assignmentPath+name)
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, nil
+	}
+
+	var assignment assignment
+	if err := entry.DecodeJSON(&assignment); err != nil {
+		return nil, err
+	}
+	return &logical.Response{
+		Data: map[string]interface{}{
+			"groups":   assignment.Groups,
+			"entities": assignment.Entities,
+		},
+	}, nil
+}
+
+// pathOIDCDeleteAssignment is used to delete an assignment
+func (i *IdentityStore) pathOIDCDeleteAssignment(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	name := d.Get("name").(string)
+	err := req.Storage.Delete(ctx, assignmentPath+name)
```

Comment:
Should we allow the deletion of an assignment if it's being used by client(s)?

Reply:
Good point. Probably not, cc @austingebauer

Reply:
I know we briefly discussed using IDs to reference other resources instead of just the name. I think preventing the deletion of resources that are currently being referenced would allow us to keep using the name--which would keep us consistent with what vault typically does.

Reply:
Yep, I agree @fairclothjm. We should prevent deletion if it's being used by any client.

Reply:
I will make this change in the Client API PR: https://github.com/hashicorp/vault/pull/12272 since this depends on both the Assignment and Client code.

Reply:
Sounds good to me 👍